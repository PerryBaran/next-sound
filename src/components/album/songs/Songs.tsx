import css from './songs.module.css'
import SongMenu from '../../songMenu/SongMenu'

interface Props {
  songName: string
  audio: string
  image?: string
  artistName: string
  albumName: string
}

export default function Songs(props: Props) {
  return (
    <li className={css.container}>
      <h4>{props.songName}</h4>
      <SongMenu songs={[props]} />
    </li>
  )
}
