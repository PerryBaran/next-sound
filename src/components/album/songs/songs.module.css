.container {
  list-style: none;
  display: flex;
  justify-content: space-between;
  margin-top: 5px;
}

.container > h4 {
}
