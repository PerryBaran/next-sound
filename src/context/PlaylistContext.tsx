'use client'

import { createContext, useContext, useState } from 'react'

interface Song {
  image?: string
  artistName: string
  albumName: string
  songName: string
  audio: string
}

interface Playlist extends Song {
  key: string
}

const PlaylistContext = createContext({
  playlist: [] as Playlist[],
  playlistIndex: 0,
  skipSong: (i: number | boolean) => {},
  handleAddToPlaylist: (
    songs: Song[],
    addNext: boolean,
    playNow: boolean
  ) => {},
  removeFromPlaylist: (i: number) => {},
  playing: false,
  handlePlaying: (isPlaying: boolean | undefined) => {}
})

const usePlaylistContext = () => useContext(PlaylistContext)

function PlaylistProvider({ children }: { children: React.ReactNode }) {
  const [playlist, setPlaylist] = useState<Playlist[]>([])
  const [playlistIndex, setPlaylistIndex] = useState(0)
  const [playing, setPlaying] = useState(false)

  const addToPlaylist = (songs: Song[], addNext = false) => {
    const data = songs.map((song) => {
      return { ...song, key: crypto.randomUUID() }
    })

    setPlaylist((prev) => {
      if (addNext) {
        const clone = [...prev]
        clone.splice(playlistIndex + 1, 0, ...data)
        return clone
      }
      return [...prev, ...data]
    })
  }

  const handlePlayNow = () => {
    if (playlist.length !== 0) {
      setPlaylistIndex((prev) => prev + 1)
    }
    setPlaying(true)
  }

  const handleAddToPlaylist = (
    songs: Song[],
    addNext: boolean,
    playNow: boolean
  ) => {
    addToPlaylist(songs, addNext)
    if (playNow) {
      handlePlayNow()
    }
  }

  const removeFromPlaylist = (i: number) => {
    setPlaylist((prev) => {
      const clone = [...prev]
      clone.splice(i, 1)
      return clone
    })

    setPlaylistIndex((prev) => {
      if (i > prev || prev === 0) {
        return prev
      }
      return prev - 1
    })
  }

  const skipSong = (value: number | boolean) => {
    const lastPlaylistIndex = playlist.length - 1
    if (typeof value === 'number') {
      if (value > lastPlaylistIndex) {
        setPlaylistIndex(lastPlaylistIndex)
      } else if (value < 0) {
        setPlaylistIndex(0)
      } else {
        setPlaylistIndex(value)
      }
    } else if (value) {
      if (playlistIndex === lastPlaylistIndex) {
        setPlaylistIndex(0)
      } else {
        setPlaylistIndex((prev) => prev + 1)
      }
    } else {
      if (playlistIndex > 0) {
        setPlaylistIndex((prev) => prev - 1)
      } else {
        setPlaylistIndex(lastPlaylistIndex)
      }
    }
  }

  const handlePlaying = (isPlaying: boolean | undefined) => {
    if (isPlaying === undefined) {
      setPlaying((prev) => !prev)
    } else {
      setPlaying(isPlaying)
    }
  }

  return (
    <PlaylistContext.Provider
      value={{
        playlist,
        playlistIndex,
        handleAddToPlaylist,
        removeFromPlaylist,
        skipSong,
        playing,
        handlePlaying
      }}
    >
      {children}
    </PlaylistContext.Provider>
  )
}

export { usePlaylistContext, PlaylistProvider }
