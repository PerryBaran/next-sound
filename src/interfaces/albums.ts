interface Songs {
  name: string
  url: string
  id: string
  createdAt: string
}

interface Users {
  name: string
  id: string
  createdAt: string
}

export default interface Albums {
  name: string
  url: string
  id: string
  User: Users
  Songs: Songs[]
  UserId: string
  createdAt: string
}
