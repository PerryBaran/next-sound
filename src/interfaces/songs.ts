interface User {
  id: string
  name: string
  email: string
  createdAt: string
}

interface Album {
  id: string
  name: string
  url: string
  user: User
  createdAt: string
}

export default interface Songs {
  id: string
  name: string
  url: string
  Albums: Album
  AlbumId: string
  createdAt: string
}
