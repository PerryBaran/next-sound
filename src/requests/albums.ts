import {
  postRequest,
  getRequest,
  patchRequest,
  deleteRequest,
  getByIdRequest
} from "./helpers"

export async function postAlbums(data: { name: string; image?: File }) {
  const response = await postRequest("albums", data)
  return response
}

export async function getAlbums(query?: {
  name?: string
  exact?: boolean
  limit?: number
}) {
  const response = await getRequest("albums", query)
  return response
}

export async function getAlbumById(id: string) {
  const response = await getByIdRequest("albums", id)
  return response
}

export async function patchAlbum(
  id: string,
  data: { name?: string; image?: File }
) {
  const response = await patchRequest("albums", id, data)
  return response
}

export async function deleteAlbum(id: string) {
  const response = await deleteRequest("albums", id)
  return response
}
