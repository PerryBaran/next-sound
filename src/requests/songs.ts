import { postRequest, getRequest, patchRequest, deleteRequest } from "./helpers/requestHelpers"

export async function postSongs(data: {
  name: string
  audio: File
  position: number
  AlbumId: string
}) {
  const response = await postRequest("songs", data)
  return response
}

export async function getSongs(query?: {
  name?: string
  exact?: boolean
  limit?: number
}) {
  const response = await getRequest("songs", query)
  return response
}

export async function patchSong(
  id: string,
  data: { name?: string; audio?: File; position?: number }
) {
  const response = await patchRequest("songs", id, data)
  return response
}

export async function deleteSong(id: string) {
  const response = await deleteRequest("songs", id)
  return response
}
