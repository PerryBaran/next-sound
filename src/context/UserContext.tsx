'use client'

import { createContext, useContext, useState, useEffect } from 'react'
import { getUserById } from '@/requests/users'
import Cookie from 'js-cookie'
import jwtDecode from 'jwt-decode'

interface User {
  name?: string
  id?: string
}

const emptyUser = {
  name: '',
  id: ''
}

const UserContext = createContext({
  user: emptyUser,
  handleLogin: (data: User | undefined) => {}
})

const useUserContext = () => useContext(UserContext)

function UserProvider({ children }: { children: React.ReactNode }) {
  const [user, setUser] = useState(emptyUser)

  const handleLogin = (data: User | undefined) => {
    if (data === undefined) {
      Cookie.remove('userToken')
      setUser(emptyUser)
    } else {
      setUser((prev) => {
        return { ...prev, ...data }
      })
    }
  }

  useEffect(() => {
    ;(async () => {
      const token = Cookie.get('userToken')
      if (!token) return
      const { id }: { id: string } = jwtDecode(token)
      const response = await getUserById(id)
      handleLogin({
        name: response.name,
        id
      })
    })()
  }, [])

  return (
    <UserContext.Provider value={{ user, handleLogin }}>
      {children}
    </UserContext.Provider>
  )
}

export { useUserContext, UserProvider }
