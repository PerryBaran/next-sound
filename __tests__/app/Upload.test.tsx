import { fireEvent, render, screen } from "@testing-library/react"
import { createContext, useContext } from "react"
import UploadAlbum from "../../src/app/upload/page"
import * as navigation from "next/navigation"
import * as UserContext from "../../src/context/UserContext"
import * as albumRequests from "../../src/requests/albums"
import * as songRequests from "../../src/requests/songs"

describe("Upload", () => {
  let mockPush = jest.fn()
  const mockName = "mock-name"

  afterEach(() => {
    jest.clearAllMocks()
  })

  beforeEach(() => {
    mockPush = jest.fn()

    jest.spyOn(navigation, "useRouter").mockReturnValue({
      back: jest.fn(),
      forward: jest.fn(),
      push: mockPush,
      refresh: jest.fn(),
      replace: jest.fn(),
      prefetch: jest.fn()
    })

    const mockContext = createContext({
      user: { name: mockName, id: "" },
      handleLogin: jest.fn()
    })
    jest
      .spyOn(UserContext, "useUserContext")
      .mockImplementation(() => useContext(mockContext))
  })

  test("renders correctly", () => {
    render(<UploadAlbum />)
    const nameInputs = screen.getAllByLabelText(/name/i)

    expect(screen.getByText(/create album/i)).toBeInstanceOf(HTMLHeadingElement)
    expect(nameInputs).toHaveLength(2)
    nameInputs.forEach((input) => {
      expect(input).toHaveAttribute("type", "text")
    })
    expect(screen.getByLabelText(/cover art/i)).toHaveAttribute("type", "file")
    expect(screen.getByText(/upload songs/i)).toBeInstanceOf(HTMLHeadingElement)
    expect(screen.getByText(/song 1/i)).toBeInstanceOf(HTMLHeadingElement)
    expect(screen.getByLabelText(/audio/i)).toHaveAttribute("type", "file")
    expect(screen.getByText(/x/i)).toBeInstanceOf(HTMLButtonElement)
    expect(screen.getByText("+")).toBeInstanceOf(HTMLButtonElement)
    expect(screen.getByText(/upload album/i)).toHaveAttribute("type", "submit")
  })

  test("remove button", () => {
    render(<UploadAlbum />)

    expect(screen.getAllByLabelText(/name/i)).toHaveLength(2)

    fireEvent.click(screen.getByText(/x/i))

    expect(screen.getAllByLabelText(/name/i)).toHaveLength(1)
  })

  test("add button", () => {
    render(<UploadAlbum />)

    expect(screen.getAllByLabelText(/audio/i)).toHaveLength(1)

    fireEvent.click(screen.getByText("+"))

    expect(screen.getAllByLabelText(/audio/i)).toHaveLength(2)
  })

  describe("submit", () => {
    const mockId = "mock-id"
    let postAlbumSpy: jest.SpyInstance
    let postSongsSpy: jest.SpyInstance

    beforeEach(() => {
      postAlbumSpy = jest.spyOn(albumRequests, "postAlbums")
      postSongsSpy = jest.spyOn(songRequests, "postSongs")
    })

    xtest("album doesn't contain a name", () => {
      render(<UploadAlbum />)

      fireEvent.click(screen.getByText(/upload album/i))

      expect(screen.getByText(/must provide a name/i)).toBeInTheDocument()
    })

    xtest("album contains no songs", () => {
      render(<UploadAlbum />)
      fireEvent.change(screen.getAllByLabelText(/name/i)[0], {
        target: { value: "album-name" }
      })
      fireEvent.click(screen.getByText(/x/i))

      fireEvent.click(screen.getByText(/upload album/i))

      expect(
        screen.getByText(/album must contain atleast one song/i)
      ).toBeInTheDocument()
    })

    xtest("atleast one song doesn't contain a name", () => {
      render(<UploadAlbum />)
      fireEvent.change(screen.getAllByLabelText(/name/i)[0], {
        target: { value: "album-name" }
      })

      fireEvent.click(screen.getByText(/upload album/i))

      expect(
        screen.getByText(/all songs must contain a name/i)
      ).toBeInTheDocument()
    })

    xtest("atleast one song doesn't contain audio", () => {
      render(<UploadAlbum />)
      fireEvent.change(screen.getAllByLabelText(/name/i)[0], {
        target: { value: "album-name" }
      })
      fireEvent.change(screen.getAllByLabelText(/name/i)[1], {
        target: { value: "song-name" }
      })

      fireEvent.click(screen.getByText(/upload album/i))

      expect(
        screen.getByText(/all songs must contain audio/i)
      ).toBeInTheDocument()
    })

    test("all data is correct", () => {
      render(<UploadAlbum />)
      postAlbumSpy.mockResolvedValue({ id: mockId })

      const albumName = "album-name"
      fireEvent.change(screen.getAllByLabelText(/name/i)[0], {
        target: { value: albumName }
      })
      const songData = {
        name: "song-name",
        audio: new File([], "audio"),
        position: 0
      }
      fireEvent.change(screen.getAllByLabelText(/name/i)[1], {
        target: { value: songData.name }
      })
      fireEvent.change(screen.getByLabelText(/audio/i), { target: { files: [songData.audio] }})
    
      fireEvent.click(screen.getByText(/upload album/i))
      
      expect(postAlbumSpy).toBeCalledWith({ name: "album-name" })
      expect(postSongsSpy).toBeCalledWith({ ...songData, AlbumId: mockId})
    })
  })
})
