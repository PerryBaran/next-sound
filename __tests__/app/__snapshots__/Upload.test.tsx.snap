// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Upload snapshot 1`] = `
<DocumentFragment>
  <div>
    <form>
      <h2>
        Create Album
      </h2>
      <label
        for="name"
      >
        <span>
          Name
        </span>
        <input
          id="name"
          name="name"
          type="text"
          value=""
        />
      </label>
      <label
        for="image"
      >
        <span>
          Cover Art
        </span>
        <input
          id="image"
          name="image"
          type="file"
        />
      </label>
      <div>
        <h2>
          Upload Songs
        </h2>
        <div
          draggable="true"
        >
          <h3>
            Song 1
          </h3>
          <label
            for="name0"
          >
            <span>
              Name
            </span>
            <input
              id="name0"
              name="name"
              type="text"
              value=""
            />
          </label>
          <label
            for="audio0"
          >
            <span>
              Audio
            </span>
            <input
              id="audio0"
              name="audio"
              type="file"
            />
          </label>
          <button
            type="button"
          >
            x
          </button>
        </div>
        <button
          type="button"
        >
          +
        </button>
        <button
          type="submit"
        >
          Upload Album
        </button>
      </div>
    </form>
  </div>
</DocumentFragment>
`;
